core/server/master/src/main/java/alluxio/master/throttle/SystemMonitor.java

```diff
@@ -0,0 +1,391 @@
+  private long mCurrentThresholdTimeIn42Sec;
+  private ServerIndicator mPitThresholdActive;
+  private ServerIndicator mAggregateThresholdActive;
+  private ServerIndicator mPitThresholdStressed;
+  private ServerIndicator mAggregateThresholdStressed;
+  private ServerIndicator mPitThresholdOverloaded;
+  private ServerIndicator mAggregateThresholdOverloaded;
+
+  private SystemStatus mCurrentSystemStatus;
+
+  private PitInfo mPrevPitInfo;
+
+  /**
+   * Pit information.
+   */
+  private static class PitInfo {
+    private SystemStatus mSystemStatus;
+    private long mPitTime;
+    private long mJVMPauseTime;
+
+    /**
+     * Constructor.
+     *
+     * @param systemStatus the system status
+     * @param pitTime the pit time
+     * @param jvmPauseTime the JVM pause time
+     */
+    public PitInfo(SystemStatus systemStatus, long pitTime, long jvmPauseTime) {
+      mSystemStatus = systemStatus;
+      mPitTime = pitTime;
+      mJVMPauseTime = jvmPauseTime;
+    }
+
+    /**
+     * @return the pit time
+     */
+    public long getPitTime() {
+      return mPitTime;
+    }
+
+    /**
+     * @return the JVP pause time
+     */
+    public long getJVMPauseTime() {
+      return mJVMPauseTime;
+    }
+
+    /**
+     * @return the system status
+     */
+    public SystemStatus getSystemStatus() {
+      return mSystemStatus;
+    }
+
+    /**
+     * Sts the system status.
+     *
+     * @param systemStatus the system status
+     */
+    public void setSystemStatus(SystemStatus systemStatus) {
+      mSystemStatus = systemStatus;
+    }
+
+    /**
+     * Sets the JVM pause time.
+     *
+     * @param jvmPauseTime the JVM pause time
+     */
+    public void setJVMPauseTime(long jvmPauseTime) {
+      mJVMPauseTime = jvmPauseTime;
+    }
+
+    /**
+     * Sets the pit time.
+     *
+     * @param pitTime the pit time
+     */
+    public void setPitTime(long pitTime) {
+      mPitTime = pitTime;
+    }
+  }
+
+  /**
+   * The system status.
+   */
+  public enum SystemStatus {
+    IDLE,
+    ACTIVE,
+    STRESSED,
+    OVERLOADED,
+  }
+
+  /**
+   * The status transition.
+   */
+  public enum StatusTransition {
+    UNCHANGED,
+    ESCALATE,
+    DEESCALATE,
+  }
+
+  /**
+   * The system monitor constructor.
+   *
+   * @param masterProcess the master process
+   */
+  public SystemMonitor(MasterProcess masterProcess) {
+    mServerIndicatorsList = new LinkedList<>();
+    mCurrentSystemStatus = SystemStatus.IDLE;
+
+    // Configurable
+    mMaxNumberOfSnapshot = Configuration.getInt(PropertyKey.MASTER_THROTTLE_OBSERVED_PIT_NUMBER);
+
+    // covert the percentage of memory usage to usedHeap
+    mCurrentThresholdTimeIn42Sec = 0;
+    reInitTheThresholds();
+    mPrevPitInfo = new PitInfo(mCurrentSystemStatus,
+        ServerIndicator.getSystemTotalJVMPauseTime(), System.currentTimeMillis());
+
+    MetricsSystem.registerGaugeIfAbsent(
+        MetricsSystem.getMetricName("system.status"),
+        () -> mCurrentSystemStatus);
```

Comment:
this variable probably needs to be volatile

Reply:
add volatile